In [2]:
import time
import tracemalloc

# Time Measurement

Initial solution

In [75]:

import initial_solution

test = 0

for _ in range(100):

    start = time.time()

    initial_solution.board_generation()
    initial_solution.board_generation()
    initial_solution.board_generation()
    initial_solution.board_generation()
    initial_solution.board_generation()
    initial_solution.board_generation()
    initial_solution.board_generation()
    initial_solution.board_generation()
    initial_solution.board_generation()
    initial_solution.board_generation()


    test += time.time() - start

print(test)



0.013950586318969727


'\nResults:\n    1 board:\n        0.014737129211425781\n        0.0152740478515625\n        0.01544046401977539\n        0.016042709350585938\n        0.015568733215332031\n    2 boards:\n        0.0324854850769043\n        0.02798175811767578\n        0.02819085121154785\n    10 boards:\n        0.13289928436279297\n        0.1393883228302002\n        0.14920473098754883\n        0.14388561248779297\n    100 boards:\n        1.3554160594940186\n        1.3143031597137451\n\n'

Results:

    1 board:
        0.014737129211425781
        0.0152740478515625
        0.01544046401977539
        0.016042709350585938
        0.015568733215332031
    2 boards:
        0.0324854850769043
        0.02798175811767578
        0.02819085121154785
    10 boards:
        0.13289928436279297
        0.1393883228302002
        0.14920473098754883
        0.14388561248779297
    100 boards:
        1.3554160594940186
        1.3143031597137451

AI solution

In [70]:
import ai_code

test = 0

for _ in range(100):

    start = time.time()

    ai_code.board_generation()
    ai_code.board_generation()
    ai_code.board_generation()
    ai_code.board_generation()
    ai_code.board_generation()
    ai_code.board_generation()
    ai_code.board_generation()
    ai_code.board_generation()
    ai_code.board_generation()
    ai_code.board_generation()

    test += time.time() - start

print(test)


0.01638054847717285


'\nResults:\n    1 board:\n        0.01520538330078125\n        0.015534162521362305\n        0.015062332153320312\n        0.015755653381347656\n        0.014878034591674805\n    2 boards:\n        0.02589130401611328\n        0.02821946144104004\n        0.02628302574157715\n        0.029273033142089844\n        0.02648138999938965\n        0.03173327445983887\n    10 boards:\n        0.1285572052001953\n        0.13258671760559082\n        0.1498093605041504\n        0.1460113525390625\n        0.14448976516723633\n    100 boards:\n        1.3102154731750488\n        1.3037786483764648\n        1.3114650249481201\n        1.2840576171875\n        1.3036134243011475\n        1.300112247467041\n\n'

Results:

    1 board:
        0.01520538330078125
        0.015534162521362305
        0.015062332153320312
        0.015755653381347656
        0.014878034591674805
        0.01638054847717285
    2 boards:
        0.02589130401611328
        0.02821946144104004
        0.02628302574157715
        0.029273033142089844
        0.02648138999938965
        0.03173327445983887
    10 boards:
        0.1285572052001953
        0.13258671760559082
        0.1498093605041504
        0.1460113525390625
        0.14448976516723633
    100 boards:
        1.3102154731750488
        1.3037786483764648
        1.3114650249481201
        1.2840576171875
        1.3036134243011475
        1.300112247467041

In [106]:
import initial_solution
import ai_code


test_initial = 0
test_ai = 0

for _ in range(100):

    start = time.time()

    initial_solution.winning_combination([['w', 'g', 'g', 'w'], [0, 0, 0, 0], [0, 'g', 'w', 'g'], \
['g', 'w', 'w', 'w'], [0, 0, 0, 'g'], [0, 0, 0, 0], [0, 0, 0, 'w'], [0, 0, 0, 0], \
[0, 0, 'w', 'w'], ['w', 'g', 'w', 'g'], [0, 0, 0, 'w'], [0, 0, 0, 'g'], [0, 0, 'g', 'w'], \
[0, 0, 0, 'w'], [0, 0, 'g', 'g'], [0, 0, 0, 'g']])

    test_initial += time.time() - start

    start = time.time()

    ai_code.winning_combination([['w', 'g', 'g', 'w'], [0, 0, 0, 0], [0, 'g', 'w', 'g'], \
['g', 'w', 'w', 'w'], [0, 0, 0, 'g'], [0, 0, 0, 0], [0, 0, 0, 'w'], [0, 0, 0, 0], \
[0, 0, 'w', 'w'], ['w', 'g', 'w', 'g'], [0, 0, 0, 'w'], [0, 0, 0, 'g'], [0, 0, 'g', 'w'], \
[0, 0, 0, 'w'], [0, 0, 'g', 'g'], [0, 0, 0, 'g']])

    test_ai += time.time() - start

print("Initial: " + str(test_initial))
print("AI:      " + str(test_ai))



Initial: 0.028400897979736328
AI:      0.028868675231933594


Results:

1)    winning_combination([[0, 0, 0, 'w'], [0, 'g', 'g', 'w'], [0, 0, 'w', 'w'], \
    [0, 'g', 'w', 'w'], [0, 0, 0, 'g'], [0, 0, 0, 0], [0, 0, 'w', 'g'], [0, 0, 0, 'g'], [0, 0, 0, 0], \
    [0, 'w', 'w', 'w'], ['g', 'w', 'w', 'g'], [0, 0, 0, 0], [0, 0, 0, 'g'], [0, 0, 'g', 'g'], \
    ['g', 'g', 'w', 'w'], [0, 0, 'g', 'w']])

    Initial: 0.03606820106506348
    AI:      0.035883188247680664

2)    winning_combination([['w', 'w', 'w', 'w'], [0, 'g', 'g', 'w'], [0, 0, 'w', 'w'], \
[0, 'g', 'w', 'w'], [0, 0, 0, 'g'], [0, 0, 0, 0], [0, 0, 'w', 'g'], [0, 0, 0, 'g'], [0, 0, 0, 0], \
[0, 'w', 'w', 'w'], ['g', 'w', 'w', 'g'], [0, 0, 0, 0], [0, 0, 0, 'g'], [0, 0, 'g', 'g'], \
['g', 'g', 'w', 'w'], [0, 0, 'g', 'w']])

    Initial: 0.033194541931152344
    AI:      0.033440351486206055

3)    winning_combination([['w', 'g', 'g', 'w'], [0, 0, 0, 0], [0, 'g', 'w', 'g'], \
['g', 'w', 'w', 'w'], [0, 0, 0, 'g'], [0, 0, 0, 0], [0, 0, 0, 'w'], [0, 0, 0, 0], \
[0, 0, 'w', 'w'], ['w', 'g', 'w', 'g'], [0, 0, 0, 'w'], [0, 0, 0, 'g'], [0, 0, 'g', 'w'], \
[0, 0, 0, 'w'], [0, 0, 'g', 'g'], [0, 0, 0, 'g']])

    (False)

    Initial: 0.02755260467529297
    AI:      0.027727842330932617

# Memory measurement

Initial solution

In [117]:
import initial_solution

tracemalloc.start()

for _ in range(100):

    initial_solution.board_generation()

print(tracemalloc.get_traced_memory())
tracemalloc.stop()


(3266, 14329)


Results:

    For 1:
    (2630, 13963)
    (2649, 13963)
    (1951, 13963)
    (2649, 13963)
    (2649, 13963)
    (2449, 13963)
    (3266, 14329)

AI solution

In [123]:
import ai_code

tracemalloc.start()

for _ in range(100):

    ai_code.board_generation()

print(tracemalloc.get_traced_memory())
tracemalloc.stop()

(2753, 13963)


Results:

    For 1:
    (2937, 14000)
    (2704, 13963)
    (1951, 13963)
    (2992, 14055)
    (2504, 13963)
    (2753, 13963)

In [157]:
import initial_solution

tracemalloc.start()

for _ in range(100):

    initial_solution.winning_combination([['w', 'g', 'g', 'w'], [0, 0, 0, 0], [0, 'g', 'w', 'g'], \
['g', 'w', 'w', 'w'], [0, 0, 0, 'g'], [0, 0, 0, 0], [0, 0, 0, 'w'], [0, 0, 0, 0], \
[0, 0, 'w', 'w'], ['w', 'g', 'w', 'g'], [0, 0, 0, 'w'], [0, 0, 0, 'g'], [0, 0, 'g', 'w'], \
[0, 0, 0, 'w'], [0, 0, 'g', 'g'], [0, 0, 0, 'g']])

print("Initial: ")
print(tracemalloc.get_traced_memory())
tracemalloc.stop()

import ai_code

tracemalloc.start()

for _ in range(100):

    ai_code.winning_combination([['w', 'g', 'g', 'w'], [0, 0, 0, 0], [0, 'g', 'w', 'g'], \
['g', 'w', 'w', 'w'], [0, 0, 0, 'g'], [0, 0, 0, 0], [0, 0, 0, 'w'], [0, 0, 0, 0], \
[0, 0, 'w', 'w'], ['w', 'g', 'w', 'g'], [0, 0, 0, 'w'], [0, 0, 0, 'g'], [0, 0, 'g', 'w'], \
[0, 0, 0, 'w'], [0, 0, 'g', 'g'], [0, 0, 0, 'g']])

print("AI: ")
print(tracemalloc.get_traced_memory())
tracemalloc.stop()

Initial: 
(3747, 17454)
AI: 
(2129, 17454)


Results:

1)    winning_combination([[0, 0, 0, 'w'], [0, 'g', 'g', 'w'], [0, 0, 'w', 'w'], \
    [0, 'g', 'w', 'w'], [0, 0, 0, 'g'], [0, 0, 0, 0], [0, 0, 'w', 'g'], [0, 0, 0, 'g'], [0, 0, 0, 0], \
    [0, 'w', 'w', 'w'], ['g', 'w', 'w', 'g'], [0, 0, 0, 0], [0, 0, 0, 'g'], [0, 0, 'g', 'g'], \
    ['g', 'g', 'w', 'w'], [0, 0, 'g', 'w']])

    Initial: 
    (4065, 18222)
    AI: 
    (1460, 18222)

2)    winning_combination([['w', 'w', 'w', 'w'], [0, 'g', 'g', 'w'], [0, 0, 'w', 'w'], \
    [0, 'g', 'w', 'w'], [0, 0, 0, 'g'], [0, 0, 0, 0], [0, 0, 'w', 'g'], [0, 0, 0, 'g'], [0, 0, 0, 0], \
    [0, 'w', 'w', 'w'], ['g', 'w', 'w', 'g'], [0, 0, 0, 0], [0, 0, 0, 'g'], [0, 0, 'g', 'g'], \
    ['g', 'g', 'w', 'w'], [0, 0, 'g', 'w']])

    Initial: 
    (3884, 18225)
    AI: 
    (1515, 18225)

3)    winning_combination([['w', 'g', 'g', 'w'], [0, 0, 0, 0], [0, 'g', 'w', 'g'], \
    ['g', 'w', 'w', 'w'], [0, 0, 0, 'g'], [0, 0, 0, 0], [0, 0, 0, 'w'], [0, 0, 0, 0], \
    [0, 0, 'w', 'w'], ['w', 'g', 'w', 'g'], [0, 0, 0, 'w'], [0, 0, 0, 'g'], [0, 0, 'g', 'w'], \
    [0, 0, 0, 'w'], [0, 0, 'g', 'g'], [0, 0, 0, 'g']])

    (False)

    Initial: 
    (3747, 17454)
    AI: 
    (2129, 17454)

# Висновок

BlackBox впринципі доволі добре оптимізував гру Eternas. В цілому він змінив небагато, в функції board_generation взагалі зробив тільки 1 табуляцію рядка, а в winning_combination забрав пару змінних і розставив іфки в зручнішому порядку. І це, як показують результати аналізу, доволі добре пришвидшило швидкість виконання завдань і займає менше місця.
Оскільки у грі використовується модуль random то для точніших обчислень була потрібна відповідно доволі велика вибірка. Аналіз показав, що, хоч і швидкість була доволі однакова, частіше іменно функції підправлені штучним інтелектом були швидші (більшість запусків показували таке). Щодо використання памяті, то в board_generation все як зажди було рандомно і результати, відповідно, майже одинакові. А от в winning_combination ші показує свої правки набагато краще. Пам'ять його функція займає чуть не завжди в 2 рази менше.
З PEP8 проблем ніяких не бачив, може тому що його зміни були не глобальні.
Але він не згенерив то що я просив з 1 разу, хоя і вказав і умову і розв'язок і конкретно розписав що від нього хочу. Як і інколи чатгпт, він жостко тупив, поки я нарешті промптами я його не спрямував у відповідне русло.